File: solutions/haskell/two-bucket/5/src/TwoBucket.hs
{-# LANGUAGE BlockArguments #-}
{-# LANGUAGE LambdaCase #-}
{-# LANGUAGE TemplateHaskell #-}

module TwoBucket (measure) where

import Control.Lens (ASetter', use, (%=), (.=), (|>))
import Control.Lens.TH (makeClassy_)
import Control.Monad (when)
import Control.Monad.Except (MonadError, runExcept, throwError)
import Control.Monad.Reader (ask, runReaderT)
import Control.Monad.State (MonadState, evalStateT)
import Data.Function ((&))
import Data.Sequence (Seq ((:<|)))
import Data.Sequence qualified as Seq
import Data.Set (Set)
import Data.Set qualified as Set

type TaskState = (Int, (Int, Int))

data AlgoState = AlgoState {visited :: Set (Int, Int), queue :: Seq TaskState}
    deriving (Show)
makeClassy_ ''AlgoState

measure :: (Int, Int) -> Int -> Maybe (Int, (Int, Int))
measure (capacity1, capacity2) target =
    evalStateT measure' initialAlgoState
        & runExcept
        & either Just (\() -> Nothing)
  where
    forbiddenBuckets = (0, capacity2)

    measure' = do
        _queue |>= (1, (capacity1, 0)) -- start with filling the first bucket
        whileJust pop \gameState@(moves, buckets@(bucket1, bucket2)) -> do
            when (target `elem` [bucket1, bucket2]) $ stopWith gameState
            _visited %= Set.insert buckets
            runReaderT (addNextMoves buckets) (moves + 1)

    add buckets = do
        v <- use _visited
        when (buckets `notElem` v && buckets /= forbiddenBuckets) do
            step <- ask
            _queue |>= (step, buckets)

    addNextMoves (bucket1, bucket2) = do
        when (bucket1 < capacity1) do
            when (bucket2 > 0) do
                let d = min (capacity1 - bucket1) bucket2
                add (bucket1 + d, bucket2 - d) -- pour snd -> fst
            add (capacity1, bucket2) -- fill fst
        when (bucket2 < capacity2) do
            when (bucket1 > 0) do
                let d = min (capacity2 - bucket2) bucket1
                add (bucket1 - d, bucket2 + d) -- pour fst -> snd
            add (bucket1, capacity2) -- fill snd
        when (bucket1 > 0) $ add (0, bucket2) -- empty fst
        when (bucket2 > 0) $ add (bucket1, 0) -- empty snd

pop :: (MonadState AlgoState m) => m (Maybe TaskState)
pop = do
    q <- use _queue
    case q of
        Seq.Empty -> pure Nothing
        s :<| q' -> do
            _queue .= q'
            pure $ Just s

whileJust :: (Monad m) => m (Maybe a) -> (a -> m ()) -> m ()
whileJust get action = loop
  where
    loop =
        get >>= \case
            Nothing -> pure ()
            Just t -> action t *> loop

initialAlgoState :: AlgoState
initialAlgoState = AlgoState{visited = Set.empty, queue = Seq.empty}

stopWith :: (MonadError e m) => e -> m a
stopWith = throwError

(|>=) :: (MonadState s m) => ASetter' s (Seq a) -> a -> m ()
l |>= a = l %= (|> a)

File: solutions/cpp/two-bucket/1/two_bucket.cpp
#include "two_bucket.h"

#include <queue>
#include <set>

namespace two_bucket {

using namespace std;

using BucketsState = pair<int, int>;

struct State {
    BucketsState buckets;
    int move_count;
};

#define add(_b1, _b2)                                       \
    do {                                                    \
        bool const stateIsForbidden =                       \
            (start_bucket == bucket_id::one                 \
                 ? _b1 == 0 and _b2 == bucket2_capacity     \
                 : _b1 == bucket1_capacity and _b2 == 0);   \
        BucketsState const bs = {_b1, _b2};                 \
        if (not visited.count(bs) and not stateIsForbidden) \
            q.push(State{bs, move_count + 1});              \
    } while (0)

measure_result measure(int const bucket1_capacity, int const bucket2_capacity,
                       int const target_volume, bucket_id const start_bucket) {
    set<BucketsState> visited;
    queue<State> q;
    {
        State startState;
        startState.move_count = 1;
        switch (start_bucket) {
            case bucket_id::one:
                startState.buckets.first = bucket1_capacity;
                break;
            case bucket_id::two:
                startState.buckets.second = bucket2_capacity;
                break;
        }
        q.push(startState);
    }
    while (not q.empty()) {
        State const state = q.front();
        q.pop();
        int const move_count = state.move_count;
        BucketsState const buckets = state.buckets;
        int const bucket_1 = state.buckets.first;
        int const bucket_2 = state.buckets.second;

        // check if state is goal
        if (bucket_1 == target_volume) {
            return measure_result{state.move_count, bucket_id::one,
                                  state.buckets.second};
        }
        if (bucket_2 == target_volume) {
            return measure_result{state.move_count, bucket_id::two,
                                  state.buckets.first};
        }
        visited.insert(buckets);

        // build next moves
        // fill one
        if (bucket_1 < bucket1_capacity) {
            // pour two -> one
            if (bucket_2 > 0) {
                int const d = min(bucket1_capacity - bucket_1, bucket_2);
                add(bucket_1 + d, bucket_2 - d);
            }
            add(bucket1_capacity, bucket_2);
        }
        // fill two
        if (bucket_2 < bucket2_capacity) {
            // pour one -> two
            if (bucket_1 > 0) {
                int const d = min(bucket2_capacity - bucket_2, bucket_1);
                add(bucket_1 - d, bucket_2 + d);
            }
            add(bucket_1, bucket2_capacity);
        }
        // empty one
        if (bucket_1 > 0) {
            add(0, bucket_2);
        }
        // empty two
        if (bucket_2 > 0) {
            add(bucket_1, 0);
        }
    }
    throw nullptr;
}

}  // namespace two_bucket
